Add shrinking to ArrayBuffer, take 3

eed3si9n opened the pull request:
This is an updated version of #7893 / #7658, addressing my own review comments.

- Changes the name from `def shrinkToSize()` to `def trimToSize()` following https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#trimToSize--
- Fixes minor whitespaces.
- Fixes handling of buffer size near `Int.MaxValue`

### commit comment

The ArrayBuffer already had the logic to grow and expand an array into
another twice the size, when adding a new element to a full array.

This PR adds a complementary `trimToSize` method, that can shrink
the array and replace it with a smaller one, if the number of occupied
positions in the array falls below a fraction of its length.

Following suggestions in code review, we keep this as a separate
method that a programmer has to invoke directly, and not as part of
other operations.

diesalbla committed:
Add shrinking to ArrayBuffer

The ArrayBuffer already had the logic to grow and expand an array into
another twice the size, when adding a new element to a full array.

This PR adds a complementary `trimToSize` method, that can shrink
the array and replace it with a smaller one, if the number of occupied
positions in the array falls below a fraction of its length.

Following suggestions in code review, we keep this as a separate
method that a programmer has to invoke directly, and not as part of
other operations.

retronym commented:
Noting that `s.c.m.ArrayDeque` (new in 2.13) already has:

```
  /**
    * Trims the capacity of this ArrayDeque's instance to be the current size
    */
  def trimToSize(): Unit = resize(length - 1)
```

So 👍 for consistency. 

It _could_ be moved into the `Buffer` interface, documented to be a no-op for buffers where it doesn't make sense. But I don't think it is something that would be abstracted over often enough to really warrant that.

retronym commented:
That made me spot an inconsistency between `ArrayDeque` and `ArrayBuffer`:

```
  /**
    * Note: This does not actually resize the internal representation.
    * See clearAndShrink if you want to also resize internally
    */
  def clear(): Unit = {
    while(nonEmpty) {
      removeHeadAssumingNonEmpty()
    }
  }

  /**
    * Clears this buffer and shrinks to @param size
    *
    * @param size
    * @return
    */
  def clearAndShrink(size: Int = ArrayDeque.DefaultInitialSize): this.type = {
    reset(array = ArrayDeque.alloc(size), start = 0, end = 0)
    this
  }
```

eed3si9n commented:
```scala
  def clear(): Unit = reduceToSize(0)
```

It's a bit confusing but `ArrayBuffer#reduceToSize` doesn't change the buffer size while `ArrayBuffer.reduceSize` is used for reducing the buffer size.
Maybe I should change the name of `reduceSize` to `reduceBufferSize`.

I can also add clearAndShrink.

eed3si9n committed:
Refactor the ArrayBuffer imple closer to ArrayDeque

This refactors the ArrayBuffer implementation so some details are closer to ArrayDeque.
For example, this adds `clearAndShrink(size: Int): this.type`, which as the name suggest both clears and trims the buffer.
